zebra-state/src/service/non_finalized_state/chain.rs

```diff
@@ -192,10 +216,16 @@ impl UpdateWith<PreparedBlock> for Chain {
             self.update_chain_state_with(&prepared.new_outputs);
             // add the utxos this consumed
             self.update_chain_state_with(inputs);
-            // add sprout anchor and nullifiers
+
+            // add the shielded data
             self.update_chain_state_with(joinsplit_data);
-            // add sapling anchor and nullifier
-            self.update_chain_state_with(sapling_shielded_data);
+            if let Some(s) = sapling_shielded_data_per_spend_anchor {
+                self.update_chain_state_with(s);
+            }
+            if let Some(s) = sapling_shielded_data_shared_anchor {
+                self.update_chain_state_with(s);
+            }
```

Comment:
When the `Option<&Option<_>>` is fixed, these should change to:
```suggestion
            self.update_chain_state_with(s);
            self.update_chain_state_with(s);
```